File: public/lib/frame-buffer.js
class FrameBuffer {
    constructor() {
        this.fbo
        this.internalFormat
        this.format
        this.type
        this.width
        this.height
        this.linear
        this.depth
        this.depthTexture
        this.textures = []
        this.drawBuffers = []
    }
    static Make(gl, width, height, internalFormat, format, type, linear, depth) {
        let frame = new FrameBuffer()
        frame.set(width, height, internalFormat, format, type, linear, depth)
        RenderSystem.MakeFrameBuffer(gl, frame)
        return frame
    }
    set(width, height, internalFormat, format, type, linear, depth) {
        if (format.length !== internalFormat.length || format.length !== type.length) {
            throw new Error("framebuffer invalid")
        }
        this.internalFormat = internalFormat
        this.format = format
        this.type = type
        this.width = width
        this.height = height
        this.linear = linear === "linear"
        this.depth = depth === "depth"
    }
    Resize(gl, width, height) {
        this.width = width
        this.height = height
        RenderSystem.SetFrameBuffer(gl, this.fbo)
        RenderSystem.UpdateFrameBuffer(gl, this)
    }
}

File: client/graphics/frame-buffer.go
package graphics

import (
	"syscall/js"
)

// FrameBuffer struct
type FrameBuffer struct {
	Fbo            js.Value
	internalFormat []js.Value
	format         []js.Value
	typeOf         []js.Value
	Width          int
	Height         int
	linear         bool
	depth          bool
	DepthTexture   js.Value
	Textures       []js.Value
	drawBuffers    []interface{}
}

// FrameBufferInit func
func FrameBufferInit(width, height int, internalFormat, format, typeOf []js.Value, linear, depth bool) *FrameBuffer {
	if len(internalFormat) != len(format) || len(internalFormat) != len(typeOf) {
		panic("frame format lengths differ")
	}
	f := &FrameBuffer{}
	f.internalFormat = internalFormat
	f.format = format
	f.typeOf = typeOf
	f.Width = width
	f.Height = height
	f.linear = linear
	f.depth = depth
	f.Textures = make([]js.Value, len(format))
	f.drawBuffers = make([]interface{}, len(format))
	return f
}

// FrameBufferMake func
func FrameBufferMake(gl js.Value, width, height int, internalFormat, format, typeOf []js.Value, linear, depth bool) *FrameBuffer {
	f := FrameBufferInit(width, height, internalFormat, format, typeOf, linear, depth)
	RenderSystemMakeFrameBuffer(gl, f)
	return f
}

// Resize func
func (me *FrameBuffer) Resize(gl js.Value, width, height int) {
	me.Width = width
	me.Height = height
	RenderSystemSetFrameBuffer(gl, me.Fbo)
	RenderSystemUpdateFrameBuffer(gl, me)
}
